File: src/edu/uf/interactable/MIP2.java
package edu.uf.interactable;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.intracellularState.Phenotype;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Util;

public class MIP2 extends Chemokine{
    public static final String NAME = "MIP2";
    public static final int NUM_STATES = 1;
    
    private static MIP2 molecule = null;
    

    private MIP2(double[][][][] qttys, Diffuse diffuse) {
    	super(qttys, diffuse, NAME);
        Neutrophil.setChemokine(MIP2.NAME);
        this.setPhenotye(Phenotype.createPhenotype());
    }

    public static MIP2 getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new MIP2(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static MIP2 getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return Constants.Kd_MIP2;
    }
    
    /*public void turnOver(int x, int y, int z) {
    	this.pdec(1-Constants.MIP2_HALF_LIFE, 0, x, y, z);
    }*/
    
    public void degrade() {
    	degrade(Constants.MIP2_HALF_LIFE, 0);
    }

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    }

    public int getIndex(String str) {
        return 0;
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Neutrophil) {
        	Cell cell = (Cell) interactable;
        	if (!cell.isDead()) { 
        		Interactions.secrete(cell, this, Constants.N_MIP2_QTTY, x, y, z, 0);
        		Interactions.bind(cell, this, x, y, z, 0);
            }
        	return true;
        }
        
        if (interactable instanceof PneumocyteII) 
        	return Interactions.secrete((Cell) interactable, this, Constants.P_MIP2_QTTY, x, y, z, 0); 
        
        if (interactable instanceof Macrophage) return true;
        	//return Interactions.secrete((Leukocyte) interactable, this, Constants.MA_MIP2_QTTY, x, y, z, 0);
        
        return interactable.interact(this, x, y, z); 
    }

    public double chemoatract(int x, int y, int z) {
        return Util.activationFunction(this.get(0, x, y, z), Constants.Kd_MIP2, Constants.VOXEL_VOL, 1, Constants.STD_UNIT_T) + Constants.DRIFT_BIAS;
    }

	@Override
	public String getName() {
		return NAME;
	}

	@Override
	public double getThreshold() {
		return -1;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}
}

File: cpp/src/edu/uf/interactable/MIP2.cpp
/*
 * MIP2.cpp
 *
 *  Created on: Jul 19, 2024
 *      Author: henriquedeassis
 */

#include "MIP2.h"
#include "../utils/Id.h"
#include "../utils/Constants.h"
#include "../utils/Util.h"
#include "../primitives/Interactions.h"
#include "Macrophage.h"
#include "PneumocyteII.h"
#include <memory>

namespace edu {
namespace uf {
namespace interactable {

using namespace primitives;

const int MIP2::classId = Id::getId();
const string MIP2::NAME = "MIP2";
const int MIP2::NUM_STATES = 1;
MIP2* MIP2::molecule = nullptr;

MIP2::MIP2(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Chemokine(qttys, xbin, ybin, zbin, wbin, diffuse) {
    setPhenotype(Id::getId());
}

MIP2* MIP2::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
    if (molecule == nullptr) {
        double**** values = new double***[wbin];
        for (int i = 0; i < wbin; ++i) {
            values[i] = new double**[xbin];
            for (int j = 0; j < xbin; ++j) {
                values[i][j] = new double*[ybin];
                for (int k = 0; k < ybin; ++k) {
                    values[i][j][k] = new double[zbin]();
                }
            }
        }
        molecule = new MIP2(values, xbin, ybin, zbin, wbin, diffuse);
    }
    return molecule;
}

int MIP2::getClassId() const {
	return classId;
}

double MIP2::getKd() {
    return utils::constexprants::Kd_IL10;
}

void MIP2::degrade() {
    //degrade(utils::constexprants::TNF_HALF_LIFE, 0);
}

int MIP2::getIndex(const std::string& str) const {
    return 0;
}

void MIP2::computeTotalMolecule(int x, int y, int z) {
    totalMoleculesAux[0] += get(0, x, y, z);
}

bool MIP2::templateInteract(Interactable* interactable, int x, int y, int z) {
	if (dynamic_cast<Neutrophil*>(interactable)) {
		Cell* cell  = static_cast<Cell*>(interactable);
	    if(!cell->isDead()){
	    	Interactions::secrete(cell, this, utils::constexprants::N_MIP2_QTTY, x, y, z, 0);
	    	Interactions::bind(cell, this, x, y, z, 0);
	    }
	    return true;
	}
	if (dynamic_cast<PneumocyteII*>(interactable))
		return Interactions::secrete(static_cast<Cell*>(interactable), this, utils::constexprants::P_MIP2_QTTY, x, y, z, 0);
	if (dynamic_cast<Macrophage*>(interactable))
		return Interactions::secrete(static_cast<Cell*>(interactable), this, utils::constexprants::MA_MIP2_QTTY, x, y, z, 0);

    return interactable->interact(this, x, y, z);
}

string MIP2::getName() const {
    return NAME;
}

double MIP2::getThreshold() const {
    return 0;
}

int MIP2::getNumState() const {
    return NUM_STATES;
}

bool MIP2::isTime() const {
    return true;
}

double MIP2::chemoatract(int x, int y, int z){
	return utils::Util::activationFunction(this->get(0, x, y, z), utils::constexprants::Kd_MIP2, utils::constexprants::VOXEL_VOL, 1, utils::constexprants::STD_UNIT_T) + utils::constexprants::DRIFT_BIAS;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */
